src/databricks/labs/ucx/assessment/aws.py

```diff
@@ -0,0 +1,239 @@
+                continue
+            actions = action["Action"]
+            s3_action = []
+            if isinstance(actions, list):
+                for single_action in actions:
+                    if single_action in self.S3_ACTIONS:
+                        s3_action.append(single_action)
+                        continue
+            elif actions in self.S3_ACTIONS:
+                s3_action = [actions]
+
+            if not s3_action:
+                continue
+            for resource in action.get("Resource", []):
+                match = re.match(self.S3_REGEX, resource)
+                if match:
+                    yield AWSPolicyAction("s3", set(s3_action), match.group(1))
+
+    def _run_json_command(self, command: str):
+        code, output, error = self._command_runner(command)
+        if code != 0:
+            logger.error(error)
+            return None
+        return json.loads(output)
+
+
+class AWSResourcePermissions:
+    def __init__(
+        self,
+        ws: WorkspaceClient,
+        aws_resources: AWSResources,
+        folder: str | None = None,
+    ):
+        if not folder:
+            folder = f"/Users/{ws.current_user.me().user_name}/.ucx"
+        self._folder = folder
+        self._aws_resources = aws_resources
+        self._ws = ws
+        self._field_names = [_.name for _ in dataclasses.fields(AWSInstanceProfileAction)]
+
+    def _get_instance_profiles(self) -> Iterable[AWSInstanceProfile]:
+        instance_profiles = self._ws.instance_profiles.list()
+        for instance_profile in instance_profiles:
+            yield AWSInstanceProfile(instance_profile.instance_profile_arn, instance_profile.iam_role_arn)
+
+    def _get_instance_profiles_access(self):
+        instance_profiles = list(self._get_instance_profiles())
+        tasks = []
+        for instance_profile in instance_profiles:
+            tasks.append(partial(self._get_instance_profile_access_task, instance_profile))
+
+        return sum(Threads.strict("Scanning Instance Profiles", tasks), [])
+
+    def _get_instance_profile_access_task(self, instance_profile: AWSInstanceProfile):
+        return list(self._get_instance_profile_access(instance_profile))
+
+    def _get_instance_profile_access(self, instance_profile: AWSInstanceProfile):
+        policies = list(self._aws_resources.list_role_policies(instance_profile.role_name))
+        for policy in policies:
+            actions = list(self._aws_resources.get_role_policy(instance_profile.role_name, policy_name=policy))
+            for action in actions:
+                yield AWSInstanceProfileAction(
+                    instance_profile.instance_profile_arn,
+                    action.resource_type,
+                    str(action.action),
+                    action.resource_path,
+                    instance_profile.iam_role_arn,
+                )
+        attached_policies = list(self._aws_resources.list_attached_policies_in_role(instance_profile.role_name))
```

Comment:
See: you don't really need to yield from those methods, because you eagerly materialize iterators

Reply:
Fixed